iucvterm/doc/ttyrun.8
.\" ttyrun.8
.\"
.\"
.\" Copyright IBM Corp. 2010, 2017
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\" -------------------------------------------------------------------------
.TH "ttyrun" "8" "December 2011" "s390-tools" "System Management Commands"
.
.ds s ttyrun
.
.
.SH NAME
ttyrun \- Start a program if a specified terminal device is available
.
.
.
.SH SYNOPSIS
.B \*s
.RB [ \-V | \-\-verbose ]
.RB [ \-e | \-\-exitstatus
.IR status ]
.I term
.I program
.RI [ "program_options" ]
.br
.B \*s
.RB [ \-h | \-\-help ]
.br
.B \*s
.RB [ \-v | \-\-version ]
.
.
.
.SH DESCRIPTION
\fB\*s\fP is typically started during system initialization and is used
to prevent a respawn through the
.BR init (8)
program when a terminal is not available.

\fIterm\fP is the name of the terminal device and is a path relative to
the \f(CW/dev\fP directory, for example, specify \f(CWhvc0\fP for
\f(CW/dev/hvc0\fP.
.br
If the specified terminal device can be opened, \fB\*s\fP starts the
specified program.

If the terminal device cannot be opened, the behavior of \fB\*s\fP
depends on the \fB\-e\fP option:
.
.RS 2
.IP "\(bu" 2
If the \fB\-e\fP option has been specified, \fB\*s\fP exits with the
specified return value, or
.IP "\(bu" 2
If the \fB\-e\fP option has not been specified,  \fB\*s\fP sleeps until
it receives a signal that causes an exit.
.RE
.PP
\fIprogram\fP is an absolute path to the program to be started by
\fB\*s\fP and \fIprogram_options\fP specify additional arguments.
Depending on the program, arguments might be required.  The variable
\f(CW%t\fP in the \fIprogram_options\fP is resolved to the terminal
device specified with \fIterm\fP.
.
.
.
.SH OPTIONS
.TP
.BR \-e ", " \-\-exitstatus\~\fIstatus\fP
Specifies an exit status that is returned when the terminal device
is not available.  \fIstatus\fP must be an integer in the range 1 to 255.

You can use this status value in an upstart job file to prevent
respawning.
.
.TP
.BR \-V ", " \-\-verbose
Displays syslog messages.
.
.TP
.BR \-h ", " \-\-help
Displays a short help text, then exits.
.
.TP
.BR \-v ", " \-\-version
Displays the version number of \fB\*s\fP, then exits.
.
.
.
.SH "RETURN VALUES"
\fB\*s\fP exits with one of the following return values to report an
error condition:
.TP
.B 1
\fB\*s\fP has been started with an argument that is not valid or
required but missing.
.TP
.B 2
\fB\*s\fP could open the file specified for \fIterm\fP but the
file is not a terminal device.
.TP
.B 3
\fB\*s\fP could not start the specified program.
.PP
The return values 1 to 3 might also be returned when the \fB\-e\fP
option is used and the terminal device is not available.
.TP
.B 4 \- 255
The terminal device is not available and the \fB\-e\fP option
specifies an exit status in this range.
.
.
.
.SH "EXAMPLES"
.SS inittab
To start \fB/sbin/agetty\fP on terminal device "hvc1", specify:
.PP
.ft CR
.in +0.25in
.nf
h1:2345:respawn:/sbin/\*s hvc1 /sbin/agetty -L 9600 %t linux
.fi
.in -0.25in
.ft
.
.SS upstart job/event files
To start \fB/sbin/agetty\fP on terminal device "hvc1", add the following
settings to the job file:
.PP
.ft CR
.in +0.25in
.nf
respawn
normal exit 42
exec /sbin/\*s -e 42 hvc1 /sbin/agetty -L 9600 %t linux
.fi
.in -0.25in
.ft
.PP
With the normal exit statement, you specify an exit status that will
prevent upstart from respawning the program.  To prevent respawning with
\fB\*s\fP, you must specify the same value for the \fB\-e\fP option.
.
.
.
.SH "SEE ALSO"
.BR agetty (8),
.BR mingetty (8),
.BR inittab (5),
.BR events (5)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option
  - lshwc: Add support for delta counter value display
  - lspai: Add output '--format' option
  - lsreipl: Add secure boot state to output
  - lswhc: Add short names to lshwc output
  - pv_tools: Add Bash and Zsh completions
  - pvapconfig: Add '--unbind' option
  - pvimg/boot: Print error messages from stage3a bootloader
  - pvimg: Add support for CCK update
  - pvsecret: Add support for CCK update
  - pvsecret: Allow retrieving secrets by index & warn for duplicated entries
  - pvsecret: Deny adding secrets with duplicated secret IDs
  - zdev: Add support for virtio devices
  - zipl: Enhance mirror support
  - zipl: Implement '--dry-run' option for all dump jobs
  - zipl_helper.device-mapper: Support mirrors over NVMe devices
  - zkey/dracut: Add a dracut config file for zkey
  - zkey/initramfs: Update initramfs hook to correct drivers and include zkey plugins
  - zkey: Add support for converting a clear-key LUKS2 volume to use a secure key

  Bug Fixes:
  - chpstat: Add missing CMG 5 data fields
  - chpstat: Fix DPU utilization calculation
  - libutil/util_file: Handle over-read in util_file_read_fd()
  - pvattest: Fix successful 'check' evaluation
  - pvsecret: Fix some edge cases for plaintext keys
  - zipl_helper.device-mapper: Fix imprecise is_device_mapper() predicate
  - zkey: Fix EP11 secure key reencipher function
  - zpcictl: Fix command line parsing for invalid options

* __v2.37.0 (2025-02-07)__

  For Linux kernel version: 6.13

  Changes of existing tools:
  - dbginfo.sh: Add details on CPU-measurement
  - dbginfo.sh: Add new crypto command
  - dbginfo.sh: Add overview commands and crypto update
  - dbginfo.sh: Adding kdump info
  - dbginfo.sh: Removing outdated email references
  - dbginfo.sh: Rework network section
  - dbginfo.sh: Update copyright 2nd year
  - pvimg: Add '--(enable|disable)-image-encryption' flags to 'pvimg create'
  - pvimg: Add '--cck <FILE>' command line option and make '--comm-key' an alias
  - pvimg: Add '--hdr-key' command line option to 'pvimg create'
  - pvimg: Rename '--key' into '--hdr-key' and use '--key' as an alias (for 'pvimg info')
  - pvsecret: Add support for retrievable secrets
  - ziorep_config: Add PCHID field to adapter report
  - ziorep_traffic: Add DEVBUSID column to traffic report
  - ziorep_utilization: Add --fcp-device parameter to print virtual adapter report
  - ziorep_utilization: Add PCHID column to physical adapter report
  - ziorep_utilization: Now prints only physical adapter report by default
  - ziorep_utilization: Swap Bus-ID and CHPID columns in virtual adapter report
  - zipl/boot: Increase section size for eckd_mv dumper
  - zkey: Add support for listing and importing protected virtualization secrets

  Bug Fixes:
  - chpstat: Fix invalid utilization data on older kernels
  - opticsmon: Fix runaway loop in on_link_change()
  - zipl: Update inline assembly for GCC 15
  - zipl_helper.device-mapper: Add missed step in logical device resolution

* __v2.36.0 (2024-12-06)__

  For Linux kernel version: 6.12

  s390-tools: Define Rust MSRV as 1.75.0

  Add new tools / libraries:
  - cpacfinfo: Tool to provide CPACF information
  - opticsmon: Tools to monitor optical modules for directly attached PCI based NICs
  - pvimg: Rust rewrite of genprotimg

  Changes of existing tools:
  - chpstat: Add data bandwidth utilization column
  - chpstat: Add support for full CMCB
  - chpstat: Add support for new CMG types
  - dbginfo.sh: add overview commands and crypto update
  - genprotimg:
    1. genprotimg is now a symbolic link to the new tool `pvimg create`
    2. Breaking API changes in genprotimg:
       1. An existing output file is no longer silently overwritten; to revert
          to the old behavior use the `--overwrite` option.
       2. The Linux kernel component is now checked to determine whether it is a
          binary s390x kernel. To disable this behavior use the
          `--no-component-check` option.
  - hyptop: Support for structured output (json, json-seq, csv)
  - lszfcp: Add missing fallback marker for non-good fc_host port_state
  - lszfcp: Improve speed with many SCSI devices
  - pvattest: Add attestation policy check command
  - zipl: Add support of partitions of mirror md-devices

  Bug Fixes:
  - lszcrypt: Fix wrong state showing up for removed AP queue within SE guest

iucvterm/src/ttyrun.c
/*
 * ttyrun - Start a program if a specified terminal device is available
 *
 *
 * ttyrun is typically used to prevent a respawn through the init(8)
 * program when a terminal is not available.
 * ttyrun runs the specific program if the specified terminal device
 * can be opened successfully.  Otherwise the program enters a sleep or
 * exits with a specified return value.
 *
 * Example: To start /sbin/agetty on terminal device hvc1, use:
 *
 *	 h1:2345:respawn:/sbin/ttyrun hvc1 /sbin/agetty -L 9600 %t linux
 *
 * Note: %t is resolved to the terminal device "hvc1" before /sbin/agetty
 *	 is started.
 *
 * Return values:
 *	   1 - invalid argument or parameter is missing
 *	   2 - terminal does not resolve to a terminal device
 *	   3 - starting the specified program failed
 *    1..255 - terminal is not available and the return code is
 *	       specified with the -e option
 *
 * Copyright 2017 IBM Corp.
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <syslog.h>
#include <unistd.h>

#include "lib/zt_common.h"


#define TTY_ESCAPE_STR		"%t"

#define EXIT_INVALID_ARG	1
#define EXIT_NO_TERMINAL	2
#define EXIT_EXEC_FAILED	3


static const char usage[] =
"Usage: %s [-e status] <term> <program> [<program_options>]\n"
"       %s [-h|--help] [-v|--version]\n"
"\n"
"Start the program if the specified terminal device is available.\n"
"If the terminal device cannot be opened, sleep until a signal is received\n"
"that causes an exit or exit with the return value specified with status.\n"
"\n"
"-e, --exitstatus     Specifies an exit status in the range 1 to 255.\n"
"-V, --verbose        Displays syslog messages.\n"
"-h, --help           Displays this help, then exits.\n"
"-v, --version        Displays version information, then exits.\n";

static void __noreturn help_exit(const char *prg)
{
	printf(usage, prg, prg);
	exit(EXIT_SUCCESS);
}

static void __noreturn version_exit(const char *prg)
{
	printf("%s: Start a program if a terminal device is available, "
	       "version %s\n", prg, RELEASE_STRING);
	printf("Copyright IBM Corp. 2010, 2017\n");
	exit(EXIT_SUCCESS);
}

static void err_exit(const char *prg, const char *msg)
{
	fprintf(stderr, "%s: %s\n", prg, msg);
	exit(EXIT_INVALID_ARG);
}

static void wait_and_exit(void)
{
	/* sleep until a signal is received, then exit */
	pause();
	exit(EXIT_SUCCESS);
}

static const struct option prog_opts[] = {
	{ "help",	no_argument, NULL, 'h'},
	{ "version",	no_argument, NULL, 'v'},
	{ "exitstatus",	required_argument, NULL, 'e'},
	{ "verbose",	no_argument, NULL, 'V'},
	{ NULL,		no_argument, NULL,  0 },
};

int main(int argc, char *argv[])
{
	int rc, tty, i, c, index, done, term_index, verbose;
	char terminal[PATH_MAX] = "";
	unsigned long exitstatus;


	/* parse command options */
	if (argc == 1)
		err_exit(argv[0], "One or more options are required but missing");

	exitstatus = done = term_index = verbose = 0;
	while (!done) {
		c = getopt_long(argc, argv, "-hve:V", prog_opts, NULL);
		switch (c) {
		case -1:
			done = 1;
			break;
		case 1:
			/* the first non-optional argument must be the
			 * terminal device */
			if (!strncmp(optarg, "/", 1))
				strncpy(terminal, optarg, PATH_MAX - 1);
			else
				snprintf(terminal, PATH_MAX, "/dev/%s", optarg);
			terminal[PATH_MAX - 1] = 0;
			term_index = optind - 1;
			done = 1;
			break;
		case 'e':
			errno = 0;
			exitstatus = strtoul(optarg, (char **) NULL, 10);
			if (errno == ERANGE)
				err_exit(argv[0], "The exit status must be "
					"an integer in the range 1 to 255");

			if (!exitstatus || exitstatus > 255)
				err_exit(argv[0], "The exit status must be "
					 "in the range 1 to 255");
			break;
		case 'V':
			verbose = 1;
			break;
		case 'h':
			help_exit(argv[0]);
		case 'v':
			version_exit(argv[0]);
		case '?':
			fprintf(stderr, "Try %s --help for more information\n",
				argv[0]);
			exit(EXIT_INVALID_ARG);
		}
	}
	index = optind;

	/* check terminal */
	if (!strlen(terminal))
		err_exit(argv[0], "You must specify the name of "
				  "a terminal device");

	/* any program to start? */
	if (index == argc)
		err_exit(argv[0], "You must specify a program to start");

	/* open and check terminal device */
	tty = open(terminal, O_NOCTTY | O_RDONLY | O_NONBLOCK);
	if (tty == -1) {
		if (verbose) {
			openlog(argv[0], LOG_PID, LOG_DAEMON);
			syslog(LOG_INFO, "Could not open tty %s (%s)",
			       terminal, strerror(errno));
			closelog();
		}

		/* enter wait or exit */
		if (exitstatus)
			exit(exitstatus);
		wait_and_exit();
	}
	rc = !isatty(tty);
	close(tty);
	if (rc)
		exit(EXIT_NO_TERMINAL);

	/* start getty program */
	for (i = index; i < argc; i++)
		if (!strcmp(argv[i], TTY_ESCAPE_STR) && term_index)
			argv[i] = argv[term_index];
	if (execv(argv[index], argv + index))
		exit(EXIT_EXEC_FAILED);

	exit(EXIT_SUCCESS);
}
